### PR TITLE
go/stackman: Add Stackman operator

### PR DESCRIPTION
**Pull request by mslipper**
Converts Stackman into a Kube operator by adding
the following:

- The basic Operator scaffolding. Things like RBAC, the
  manifests for the Operator itself, etc.
- CRDs for the DTL, Sequencer, Gas Oracle, Clique L1,
  batch submitter, and deployer.

Metadata:

- Fixes ENG-1673

### CODE REVIEW

File: go/stackman/api/v1/cliquel1_types.go [hunk 1]
```diff
@@ -0,0 +1,937 @@
...
+  "number": "0x0",
+  "gasUsed": "0x0",
+  "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+  "baseFeePerGas": null
+}
+`
+
+const CliqueL1InitScript = `
+geth --datadir /data init /genesis/genesis.json
+echo "password" > /data/password.txt
```
Comment (optimisticben) [inline go/stackman/api/v1/cliquel1_types.go:935]: `echo -n` to prevent the newline from being used in the password.

File: go/stackman/config/default/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+- ../crd
+- ../rbac
+- ../manager
+# [WEBHOOK] To enable webhook, uncomment all the sections with [WEBHOOK] prefix including the one in
+# crd/kustomization.yaml
+#- ../webhook
+# [CERTMANAGER] To enable cert-manager, uncomment all sections with 'CERTMANAGER'. 'WEBHOOK' components are required.
+#- ../certmanager
+# [PROMETHEUS] To enable prometheus monitor, uncomment all sections with 'PROMETHEUS'.
+#- ../prometheus
```
Comment (optimisticben) [inline go/stackman/config/default/kustomization.yaml:25]: We should uncomment?

File: go/stackman/config/default/kustomization.yaml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+patchesStrategicMerge:
+# Protect the /metrics endpoint by putting it behind auth.
+# If you want your controller-manager to expose the /metrics
+# endpoint w/o any authn/z, please comment the following line.
+- manager_auth_proxy_patch.yaml
```
Comment (optimisticben) [inline go/stackman/config/default/kustomization.yaml:31]: Should be commented, I think. We don't want to authenticate metric collection inside the cluster.

File: go/stackman/controllers/cliquel1_controller.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+	ctrl.SetControllerReference(crd, cfgMap, r.Scheme)
+	return cfgMap
+}
+
+func (r *CliqueL1Reconciler) deployment(crd *stackv1.CliqueL1) *appsv1.Deployment {
+	replicas := int32(1)
+	labels := map[string]string{
+		"app": "clique",
+	}
+	image := "ethereum/client-go:v1.10.10"
```
Comment (optimisticben) [inline go/stackman/controllers/cliquel1_controller.go:135]: This image name should not be static.

**Comment by optimisticben** [comment]
Left a few comments, but a README or some other instructions for building and testing would help 🙏 
(how do I use it)

File: go/stackman/Makefile [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+GOBIN=$(shell go env GOBIN)
+endif
+
+# Setting SHELL to bash allows bash commands to be executed by recipes.
+# This is a requirement for 'setup-envtest.sh' in the test target.
+# Options are set to exit when a recipe line exits non-zero or a piped command fails.
+SHELL = /usr/bin/env bash -o pipefail
+.SHELLFLAGS = -ec
+
+all: build
```
Comment (cfromknecht) [inline go/stackman/Makefile:56]: missing `.PHONY` targets for `all` through `envtest`?

File: go/stackman/Makefile [hunk 6]
```diff
@@ -0,0 +1,201 @@
+# VERSION defines the project version for the bundle.
```
Comment (cfromknecht) [inline go/stackman/Makefile:1]: nice comments in this Makefile 👌

File: go/stackman/controllers/batchsubmitter_controller.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
+/*
```
Comment (cfromknecht) [inline go/stackman/controllers/batchsubmitter_controller.go:1]: any ideas why the syntax highlighting doesn't work on many of the controllers?

File: go/stackman/main.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:                 scheme,
+		MetricsBindAddress:     metricsAddr,
+		Port:                   9443,
+		HealthProbeBindAddress: probeAddr,
+		LeaderElection:         enableLeaderElection,
+		LeaderElectionID:       "8103f40b.optimism-stacks.net",
```
Comment (cfromknecht) [inline go/stackman/main.go:74]: where is this ID derived from?

**Review by cfromknecht — COMMENTED** [review]
Awesome work! 🤘

File: go/gas-oracle/oracle/gas_price_oracle.go [hunk 9]
```diff
@@ -196,15 +196,15 @@ func NewGasPriceOracle(cfg *Config) (*GasPriceOracle, error) {
 	address := cfg.gasPriceOracleAddress
 	contract, err := bindings.NewGasPriceOracle(address, l2Client)
 	if err != nil {
-		return nil, err
+		return nil, wrapErr(err, "error creating contract binding")
```
Comment (optimisticben) [inline go/gas-oracle/oracle/gas_price_oracle.go:199]: Do you mean to have gas_oracle changes in this PR?

File: integration-tests/actor-tests/lib/metrics.ts [hunk 10]
```diff
@@ -1,5 +1,7 @@
 import fs from 'fs'
 import client from 'prom-client'
+import http from 'http'
+import url from 'url'
```
Comment (optimisticben) [inline integration-tests/actor-tests/lib/metrics.ts:4]: Integration test changes should be a separate PR

File: go/stackman/Makefile (hunk 5 above)
Comment (mslipper) [reply go/stackman/Makefile:56]: This makefile was autogenerated by `operator-sdk` - I can add `.PHONY` targets if that's the best practice.

File: go/stackman/main.go (hunk 8 above)
Comment (mslipper) [reply go/stackman/main.go:74]: It's randomly generated by the operator-sdk. Since we're not enabling leader election, it's effectively a no-op in this case.

File: go/gas-oracle/oracle/gas_price_oracle.go (hunk 9 above)
Comment (mslipper) [reply go/gas-oracle/oracle/gas_price_oracle.go:199]: Nope will remove.

**Comment by mslipper** [comment]
This PR has been split up:

- https://github.com/ethereum-optimism/optimism/pull/1945 for the integration tests.
- https://github.com/ethereum-optimism/optimism/pull/1946 for the operator.

Feedback in this pr sans the `.PHONY` comments Conner made have been addressed in PR #1946.
